### PR TITLE
Elk

### PR DESCRIPTION
**Pull request by xaviergonzalez**
Here is the code to run quasi deer on minrnn for the BCI task

Note that I think this code needs python 3.12 for the proger scan to work, at least on sherlock.

I didn't try to install mamba, so anything involving mamba is commented out right now (on this branch).

### CODE REVIEW

File: scripts/run.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 torch.set_float32_matmul_precision('medium')  # Or 'high'
 
 
-@hydra.main(config_path="../configs/bci", config_name="baseline_gru", version_base=None)
```
Comment (ekellbuch) [inline scripts/run.py]: todo: undo

File: scripts/run.py [hunk 2]
```diff
@@ -41,7 +41,8 @@ def train(args):
           kwargs = dict()
 
         logger = WandbLogger(name=args.experiment_name,
-                             project=args.project_name, **kwargs)
```
Comment (ekellbuch) [inline scripts/run.py]: todo: undo

File: src/deep_ssm/models/__init__.py [hunk 3]
```diff
@@ -1,7 +1,6 @@
-from .bci_models import MambaDecoder, GRUDecoder, SashimiDecoder
+from .bci_models import GRUDecoder, MinRNNDecoder
 
 all_models = {
   "bci_gru" : GRUDecoder,
-  "bci_mamba": MambaDecoder,
-  "bci_sashimi": SashimiDecoder
```
Comment (ekellbuch) [inline src/deep_ssm/models/__init__.py:6]: undo

**Review by ekellbuch — CHANGES_REQUESTED** [review]
Let's undo the code deletion for mamba

File: configs/bci/minrnn_sweep.yaml
Comment (ekellbuch) [inline configs/bci/minrnn_sweep.yaml:1]: remove file: we will do this with wandb not hydra

File: requirements.txt [hunk 4]
```diff
@@ -1,15 +1,14 @@
...
 wandb
 torchtyping
-causal-conv1d
-triton==2.2.0
-mamba-ssm[causal-conv1d]  # temporary flag to install selective_scan_cuda 
+triton
 edit-distance
-pyparsing
\ No newline at end of file
+pyparsing
```
Comment (ekellbuch) [inline requirements.txt:14]: can we just move causal-conv1d and mamba-ssm to the end? so if they fail we exit

File: scripts/run.py [hunk 5]
```diff
@@ -15,7 +15,7 @@
 torch.set_float32_matmul_precision('medium')  # Or 'high'
 
 
-@hydra.main(config_path="../configs/bci", config_name="baseline_gru", version_base=None)
+@hydra.main(config_path="../configs/bci", config_name="quasi_minrnn", version_base=None)
```
Comment (ekellbuch) [inline scripts/run.py]: no need to change here

File: scripts/run.py [hunk 6]
```diff
@@ -41,7 +41,8 @@ def train(args):
           kwargs = dict()
 
         logger = WandbLogger(name=args.experiment_name,
-                             project=args.project_name, **kwargs)
+                             project=args.project_name, 
+                             entity="xavier_gonzalez", **kwargs)
```
Comment (ekellbuch) [inline scripts/run.py]: no need to set this one here

File: scripts/run.sh [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#SBATCH --error=bci_%j_%a.err
+#SBATCH --out=bci_%j_%a.out
+#SBATCH --time=04:59:59
+#SBATCH --gres=gpu:1
+#SBATCH --cpus-per-gpu=4
+#SBATCH --constraint='GPU_SKU:A100_SXM4&GPU_MEM:80GB'
+#SBATCH --mail-type=ALL
+
+export DEEP_SSM_DATA=/scratch/groups/swl1
+module load cuda/12.4
```
Comment (ekellbuch) [inline scripts/run.sh:12]: ml python/3.9.0
ml gcc/10.1.0
ml cudnn/8.9.0.131
ml cuda/12.4.0

File: scripts/run.sh [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+
+source /scratch/users/xavier18/miniconda3/bin/activate deep_ssm
```
Comment (ekellbuch) [inline scripts/run.sh]: eval "$(conda shell.bash hook)"
conda activate deep_ssm

File: scripts/sweep_run.sh
Comment (ekellbuch) [inline scripts/sweep_run.sh:1]: we will handlw this with wandb

File: src/deep_ssm/models/__init__.py
Comment (ekellbuch) [inline src/deep_ssm/models/__init__.py:1]: let's not remove mamba

File: src/deep_ssm/models/bci_models.py
Comment (ekellbuch) [inline src/deep_ssm/models/bci_models.py:1]: can we add this in another file?

File: src/deep_ssm/models/bci_models.py [hunk 9]
```diff
@@ -197,171 +449,24 @@ def forward(self, neuralInput, dayIdx):
         return seq_out
 
 
-class MambaDecoder(BaseDecoder):
+class MinRNNDecoder(BaseDecoder):
```
Comment (ekellbuch) [inline src/deep_ssm/models/bci_models.py]: do not remove mamba

**Review by ekellbuch — COMMENTED** [review]
Let's undo the code deletion for mamba

File: scripts/run.sh (hunk 7 above)
Comment (xaviergonzalez) [reply scripts/run.sh:12]: I think it was OK without these module loads (no gcc errors), etc.

Here were the errors that I did have: 

/scratch/users/xavier18/miniconda3/envs/deep_ssm/lib/python3.12/site-packages/torch/functional.py:534: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at /opt/conda/conda-bld/pytorch_1729647378361/work/aten/src/ATen/native/TensorShape.cpp:3595.)
 2   return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
 3 /scratch/users/xavier18/miniconda3/envs/deep_ssm/lib/python3.12/site-packages/lightning/fabric/plugins/environments/slurm.py:204: The `srun` command is available on your system but is not used. HINT: If your intention is to run Lightning on SLURM, prepend your python command with `srun` like so: srun python3 run.py --config-name quasi_minrnn ...
 4 GPU available: True (cuda), used: True
 5 TPU available: False, using: 0 TPU cores
 6 HPU available: False, using: 0 HPUs
 7 LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
 8 
 9   | Name     | Type          | Params | Mode
10 ---------------------------------------------------
11 0 | model    | MinRNNDecoder | 56.2 M | train
12 1 | loss_ctc | CTCLoss       | 0      | train
13 ---------------------------------------------------
14 56.2 M    Trainable params
15 0         Non-trainable params
16 56.2 M    Total params
17 224.838   Total estimated model params size (MB)
18 20        Modules in train mode
19 0         Modules in eval mode
20 Epoch 72:  46%|████▋     | 64/138 [00:32<00:37,  1.99it/s, v_num=tnze]     
21 /scratch/users/xavier18/miniconda3/envs/deep_ssm/lib/python3.12/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:424: The 'val_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=3` in the `DataLoader` to improve performance.
22 /scratch/users/xavier18/deep_ssm/src/deep_ssm/data/data_transforms.py:176: UserWarning: Using padding='same' with even kernel lengths and odd dilation may require a zero-padded copy of the input be created (Triggered internally at /opt/conda/conda-bld/pytorch_1729647378361/work/aten/src/ATen/native/Convolution.cpp:1036.)
23   return self.conv(input, weight=self.weight, groups=self.groups, padding="same")
24 /scratch/users/xavier18/miniconda3/envs/deep_ssm/lib/python3.12/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:424: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=3` in the `DataLoader` to improve performance.
25 Testing DataLoader 0: 100%|██████████| 14/14 [00:15<00:00,  0.91it/s]   
26 `Trainer.fit` stopped: `max_steps=10000` reached.
27 LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
28 /scratch/users/xavier18/miniconda3/envs/deep_ssm/lib/python3.12/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:424: The 'test_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=3` in the `DataLoader` to improve performance.
29 ───────────────────────────────────────────────────────────────────────────────────────────────────────

File: scripts/sweep_run.sh
Comment (xaviergonzalez) [reply scripts/sweep_run.sh:1]: Yes, if you could show me how we could do that, I would really appreciate this!

File: src/deep_ssm/models/bci_models.py
Comment (xaviergonzalez) [reply src/deep_ssm/models/bci_models.py:1]: I have just added minrnn.py under mixers
